zkevm-circuits/src/evm_circuit/execution/begin_tx.rs

```diff
@@ -158,13 +153,211 @@ impl<F: Field> ExecutionGadget<F> for BeginTxGadget<F> {
+                            .collect::<Vec<_>>()
+                            .as_slice(),
+                    );
+                    // nonce / nonce-prefix
+                    input_exprs.push(select::expr(
+                        tx_nonce_is_zero.expr(),
+                        128.expr(),
+                        select::expr(
+                            tx_nonce_lt_128.expr(),
+                            tx_nonce.expr(),
+                            128.expr() + tx_nonce_byte_size.byte_size(),
+                        ),
+                    ));
+                    // address
+                    input_exprs.extend_from_slice(
+                        tx_caller_address_bytes
+                            .iter()
+                            .map(Expr::expr)
+                            .collect::<Vec<_>>()
+                            .as_slice(),
+                    );
+                    // addr-prefix
+                    input_exprs.push(148.expr());
+                    // prefix
+                    input_exprs.push(select::expr(
+                        tx_nonce_lt_128.expr(),
+                        214.expr(),
+                        214.expr() + tx_nonce_byte_size.byte_size(),
+                    ));
+                    let input_exprs: [Expression<F>; MIN_RLP_LEN_CONTRACT_CREATION_INPUT + $n] =
+                        input_exprs.try_into().unwrap();
+                    let input_rlc = cb.keccak_rlc(input_exprs);
+
+                    // 4. keccak table lookup
+                    cb.keccak_table_lookup(input_rlc, input_len.expr(), output_rlc.expr());
+                });
+            };
+        }
+
+        // Cover various cases of keccak lookup guarded by the number of bytes used to
+        // represent transaction's nonce.
+        //
+        // The tx nonce in the RLP-encoding could be n ∈ [0, 8] bytes. Depending on the
+        // byte size of tx nonce, we take `n` byte from the 8-bytes nonce array.
+        keccak_lookup!(0, tx_nonce_lt_128.expr());
+        keccak_lookup!(
+            1,
+            not::expr(tx_nonce_lt_128.expr()),
+            tx_nonce_byte_size_cmp.value_equals(1usize)
+        );
+        keccak_lookup!(2, tx_nonce_byte_size_cmp.value_equals(2usize));
+        keccak_lookup!(3, tx_nonce_byte_size_cmp.value_equals(3usize));
+        keccak_lookup!(4, tx_nonce_byte_size_cmp.value_equals(4usize));
+        keccak_lookup!(5, tx_nonce_byte_size_cmp.value_equals(5usize));
+        keccak_lookup!(6, tx_nonce_byte_size_cmp.value_equals(6usize));
+        keccak_lookup!(7, tx_nonce_byte_size_cmp.value_equals(7usize));
+        keccak_lookup!(8, tx_nonce_byte_size_cmp.value_equals(8usize));
+
+        cb.condition(tx_is_create.expr(), |cb| {
+            cb.account_write(
+                call_callee_address.expr(),
+                AccountFieldTag::CodeHash,
+                phase2_code_hash.expr(),
+                0.expr(),
+                None,
```

Comment:
i am not sure whether the creation tx always success? if not we need reversion_info here.

Reply:
Yes, good catch. I am thinking we should also have a OOG test case where we can check that all changes got reverted.

Reply:
i think @DreamWuGit is working on creation related error cases?